File: scripting_languages/task_2/python/main.py
from re import search, IGNORECASE
from collections import Counter
from logging import getLogger, basicConfig, INFO, StreamHandler


class NginxLogAnalyzer:
    def __init__(self, log_file_path: str):
        self.log_file_path = log_file_path

        self.logger = getLogger(__name__)
        basicConfig(filename='result_task.log', format='%(asctime)s - [%(levelname)s] - %(name)s - (%(filename)s).%(funcName)s(%(lineno)d) - %(message)s', level=INFO, encoding='utf-8')
        console_handler = StreamHandler()
        console_handler.setLevel(INFO)
        
        self.logger.addHandler(console_handler)
        self.patterns: dict[str, str] = {
            'sql_injection': r'(union[\s+.*?select|select[\s+\*|\s+.*?\bfrom\s+\b|\s+.*?\bwhere\s+\b|load_file|outfile|into[\s+.*?file\s*])',
            'path_traversal': r'(\.\./|\\.\\|\/\.\.\/|\\.\\.\/|\/etc\/|\.\/|~\/|\.\\|\.\\.\\|\.\/\.\/|\/\.\/|\/\/)',
            'xss_attack': r'(<\s*script\s*>)',
            'csrf_attack': r'(csrf_token=)',
            'command_injection': r'(;|\|&|\$\(.*?\)|`.*?`|&&|\|\|)',
            'shell_metacharacters': r'(\||;|\*|\?|~|\^|<|>)',
            'server_side_include': r'(<!--#.*?-->)',
            'file_inclusion': r'(\/etc\/passwd|\/etc\/shadow|\/proc\/self\/environ|\/etc\/group)',
            'remote_file_inclusion': r'(http:\/\/|ftp:\/\/|file:\/\/)'
        }

    def read_log_file(self) -> list[str]:
        with open(self.log_file_path, 'r') as file:
            logs = file.readlines()
        return logs

    def analyze_logs(self, logs: list[str]) -> Counter:
        suspicious_requests = []
        ip_blacklist: list[str] = ['94.23.74.168', '112.94.184.152', '195.101.2.195', '5.135.213.197']
        ip_whitelist: list[str] = ['8.8.8.8', '0.0.0.0']

        for log in logs:
            ip = search(r'^([\d.]+)', log).group()
            request = search(r'\"(.+?)\"', log).group(1)

            is_suspicious = False
            for pattern_name, pattern in self.patterns.items():
                if search(pattern, request, IGNORECASE):
                    is_suspicious = True
                    break

            if ip not in ip_whitelist and ip in ip_blacklist:
                is_suspicious = True

            if is_suspicious:
                suspicious_requests.append(request)

        return Counter(suspicious_requests)

    def log_results(self, counter: Counter) -> None:
        self.logger.info("Top-20 suspicious requests:")
        for request, count in counter.most_common(20):
            self.logger.info(f"Request: {request} - Occurs {count} times")


def main() -> None:
    log_analyzer = NginxLogAnalyzer("access.log")
    logs = log_analyzer.read_log_file()
    suspicious_requests_counter = log_analyzer.analyze_logs(logs)
    log_analyzer.log_results(suspicious_requests_counter)


if __name__ == "__main__":
    main()

File: scripting_languages/task_2/js/main.js
const fs = require('fs');
const { createLogger, format, transports } = require('winston');


class NginxLogAnalyzer {
    constructor(logFilePath) {
        this.logFilePath = logFilePath;

        this.logger = createLogger({
            level: 'info',
            format: format.combine(
                format.timestamp(),
                format.printf(info => `${info.timestamp} - [${info.level.toUpperCase()}] - ${info.message}`)
            ),
            transports: [
                new transports.File({ filename: 'result_task.log', options: { encoding: 'utf-8' } }),
                new transports.Console()
            ]
        });

        this.patterns = {
            sql_injection: /(union[\s+.*?select|select[\s+\*|\s+.*?\bfrom\s+\b|\s+.*?\bwhere\s+\b|load_file|outfile|into[\s+.*?file\s*])/i,
            path_traversal: /(\.\.\/|\\.\\|\/\.\.\/|\\.\\.\/|\/etc\/|\.\/|~\/|\.\\|\.\\.\\|\.\/\.\/|\/\.\/|\/\/)/i,
            xss_attack: /(<\s*script\s*>)/i,
            csrf_attack: /(csrf_token=)/i,
            command_injection: /(;|\|&|\$\(.*?\)|`.*?`|&&|\|\|)/i,
            shell_metacharacters: /(\||;|\*|\?|~|\^|<|>)/i,
            server_side_include: /(<!--#.*?-->)/i,
            file_inclusion: /(\/etc\/passwd|\/etc\/shadow|\/proc\/self\/environ|\/etc\/group)/i,
            remote_file_inclusion: /(http:\/\/|ftp:\/\/|file:\/\/)/i
        };
    }

    readLogFile() {
        return fs.readFileSync(this.logFilePath, 'utf-8').split('\n');
    }

    analyzeLogs(logs) {
        const suspiciousRequests = [];
        const ipBlacklist = ['94.23.74.168', '112.94.184.152', '195.101.2.195', '5.135.213.197'];
        const ipWhitelist = ['8.8.8.8', '0.0.0.0'];

        logs.forEach(log => {
            const ipMatch = log.match(/^([\d.]+)/);
            const requestMatch = log.match(/"(.+?)"/);

            if (ipMatch && requestMatch) {
                const ip = ipMatch[1];
                const request = requestMatch[1];

                let isSuspicious = false;
                for (const [patternName, pattern] of Object.entries(this.patterns)) {
                    if (pattern.test(request)) {
                        isSuspicious = true;
                        break;
                    }
                }

                if (ipWhitelist.indexOf(ip) === -1 && ipBlacklist.indexOf(ip) !== -1) {
                    isSuspicious = true;
                }

                if (isSuspicious) {
                    suspiciousRequests.push(request);
                }
            }
        });

        return suspiciousRequests.reduce((acc, request) => {
            acc[request] = (acc[request] || 0) + 1;
            return acc;
        }, {});
    }

    logResults(counter) {
        this.logger.info('Top-20 suspicious requests:');
        const sortedRequests = Object.entries(counter).sort((a, b) => b[1] - a[1]).slice(0, 20);
        sortedRequests.forEach(([request, count]) => {
            this.logger.info(`Request: ${request} - Occurs ${count} times`);
        });
    }
}

function main() {
    const logAnalyzer = new NginxLogAnalyzer('access.log');
    const logs = logAnalyzer.readLogFile();
    const suspiciousRequestsCounter = logAnalyzer.analyzeLogs(logs);
    logAnalyzer.logResults(suspiciousRequestsCounter);
}


main();